### PR TITLE
Rework/streammanager - fix recording subframe issue, add processing queue, configurable limits memory/fps

### PR DESCRIPTION
**Pull request by pawel-soja**
Problems that occurred:
- Subframe - The SER recording has the correct resolution header, but the frame data is saved at the original resolution.
- Signal processing, sending and saving block the possibility of taking a new photo, even though it is done in a separate thread. This is very problematic when recording short videos with a lot of FPS.

What has been done:
- You can adjust the queue size for frames by setting the value in "Maximum Buffer Size (MB)" for non blocking processing.
- You can set a FPS limit for preview. This has no effect on recording.
- After enlarging the preview image, you can record this subframe to a SER file. Very useful for recording planets.

### CODE REVIEW

**Review by knro — CHANGES_REQUESTED** [review]
So I'll have to test the streaming and recording performance. Have you ran any pre/post benchmarks? Right now the recording functionality ends up in a deadlock. Try to record (only record directly from control panel, no streaming) and after recording is done, the driver is locked.

File: libs/stream/streammanager.cpp [hunk 1]
```diff
@@ -239,109 +265,164 @@ bool StreamManager::updateProperties()
...
+        if (m_framesBuffer.size() > 0)
+        {
+            uint32_t allocated = m_framesBuffer.size() * m_framesBuffer.front().frame.size() / 1024 / 1024; // MB
+            if (allocated > LimitsN[LIMITS_BUFFER_MAX].value)
+            {
+                LOG_WARN("Frame buffer is full, skipping frame...");
+                return;
+            }
+        }
+        m_framesBuffer.push_back(std::move(TimeFrame{m_FPSFast.deltaTime(), std::vector<uint8_t>{buffer, buffer + nbytes}})); // add copy to queue
```
[the remark this answers is not in the document]
Comment (pawel-soja) [reply libs/stream/streammanager.cpp:295]: Copies is faster than waiting for record or send.
The queue is absolutely necessary when writing to the hard disk is slow.
Another way may be to force the camera driver to place frames in different buffers.

With the changes, I am able to record a movie at the full speed offered by the camera (360-720MB/s). Part of the material is briefly cached in RAM (~ 2GB) before it goes to the disk.

Sorry for the rush. Take a closer look at what I broke. Currently I have a beautiful clear sky ;)

**Comment by knro** [comment]
Right now the major issue is that recording locks up the driver after it is complete. Will try to debug but please take a look.

**Comment by knro** [comment]
So I see two threads now, the asyncStreamThread

```
if (m_framesBuffer.size() == 0)
                m_framesIncoming.wait(lock);
```

And this makes sense. However, the INDI thread is stuck in the event loop:

```
/* find next */
    do
    {
        lastcb = (lastcb + 1) % ncback;
        cp     = &cback[lastcb];
    } while (!cp->in_use || !FD_ISSET(cp->fd, rfdp));

```

So something about the new code is breaking the event loop or causing some side-effect that leads to this.

**Comment by pawel-soja** [comment]
Didn't my last commit fix the problem (#124d982)?

**Comment by knro** [comment]
No, the StreamManager object is kept throughout the lifetime of the driver, so that desctructor is called when the driver exists. Now the issue is that the driver locks up AFTER recording, not streaming. 

I actually recall that I encountered this specific issue before and I got around it by using detached thread.

**Comment by pawel-soja** [comment]
When no new frame arrives, thread only waiting.
It was my intention to keep the thread active and ready to go.
Nothing should be blocked. I recorded without preview and can't reproduce the problem. Everything works.
I test on various ASI cameras, after recompiling indi and indi-3rdparty (now size of StreamManager class is diffrent).

I would like to solve the problem, can you show me how to reproduce the problem?

**Comment by knro** [comment]
Can you record SER file? Go to INDI Control Panel --> Driver --> Streaming --> Record (Duration) for example to record 1 second to a file. After that, it locks. I tried detaching the thread but that didn't help either.

**Comment by pawel-soja** [comment]
Traditional: it works for me.

What is your driver locks up effect?
Can't record the video again?
Zombie process?

2020-11-07T19:21:20: [INFO] Record Duration(millisec): 1406.11 -- Frame count: 30 
2020-11-07T19:21:20: [INFO] Ending record after 30 frames 
2020-11-07T19:21:18: [INFO] Record file is /home/pawel/indi_2020-11-07/indi_record_2020-11-07@19-21-18.ser 
2020-11-07T19:21:18: [INFO] Starting video record (Frame count): 30. 

2020-11-07T19:21:17: [INFO] Record Duration(millisec): 1399.03 -- Frame count: 30 
2020-11-07T19:21:17: [INFO] Ending record after 30 frames 
2020-11-07T19:21:16: [INFO] Record file is /home/pawel/indi_2020-11-07/indi_record_2020-11-07@19-21-16.ser 
2020-11-07T19:21:16: [INFO] Starting video record (Frame count): 30. 

2020-11-07T19:21:14: [INFO] Record Duration(millisec): 1019.53 -- Frame count: 20 
2020-11-07T19:21:14: [INFO] Ending record after 1019.53 millisecs 
2020-11-07T19:21:13: [INFO] Record file is /home/pawel/indi_2020-11-07/indi_record_2020-11-07@19-21-13.ser 
2020-11-07T19:21:13: [INFO] Starting video record (Duration): 1 secs. 

2020-11-07T19:21:12: [INFO] Record Duration(millisec): 1022.73 -- Frame count: 20 
2020-11-07T19:21:12: [INFO] Ending record after 1022.73 millisecs 
2020-11-07T19:21:11: [INFO] Record file is /home/pawel/indi_2020-11-07/indi_record_2020-11-07@19-21-11.ser 
2020-11-07T19:21:11: [INFO] Starting video record (Duration): 1 secs.

**Comment by knro** [comment]
I'm using the V4L2 driver. I was think maybe it has to do with the file descriptor getting opening in the main thread and then the async thread writes to it and then closes it. Also async thread calls IDSetSwitch at the end, but this is already protected by a Mutex so I don't think it would cause any issues.

**Comment by pawel-soja** [comment]
I think I know what's going on.
Some drivers don't like to be stopped while exposing is in progress.
Originally, the thread blocked exposing and there was no collision (and it wasn't possible to get a lot of FPS).
Give me a minute and I'll fix it.

File: libs/stream/streammanager.cpp [hunk 2]
```diff
@@ -265,6 +265,13 @@ bool StreamManager::updateProperties()
  * Binned frame must be sent from the camera driver for this to work consistentaly for all drivers.*/
 void StreamManager::newFrame(const uint8_t * buffer, uint32_t nbytes)
 {
+    // close the data stream on the same thread as the data stream
+    if (m_isRecordingAboutToClose)
+    {
+        stopRecording();
+        return;
```
Comment (pawel-soja) [inline libs/stream/streammanager.cpp:272]: Also, i added the "return" in about to close checking to avoid referencing a buffer that might be destroyed.

**Comment by knro** [comment]
Excellent!! Now it works on SER, so great job syncing that! But if I try encoder MOEG and Record OGV (Ogg Theora), it crashes when stopping recording in theora_write_frame in the main thread. It appears to be writing the "closing frame" while in the data thread, it's also in the same function theora_write_frame so the crash make sense.

what's the best way to avoid this collision?

**Comment by knro** [comment]
So it appears that synchronizing open/write/close across 2-threads is going to be challenging. I'm thinking of moving all of that into the same thread. Write now it just performs the *write* part, maybe it wouldn't be too difficult to make it handle open and close.

This way we avoid concurrent access without having to add mutexes.

**Comment by pawel-soja** [comment]
I found a lonely mutex in header so I used it ;) I tested it on ASI and Web cameras.
Performance is better. I want to thank you for your support and testing.

**Comment by knro** [comment]
Thank you!! That fixed it! Users going to love this I'm sure, thanks for all the hard work.

**Pull request merged**